File: c++/move-zeros.cpp
#include <iostream>
#include <vector>
using namespace std;

class Solution {
public:
    void moveZeroes(vector<int>& nums) {
        int i = 0;
        for(int j = 0; j < nums.size(); j++){
            if(nums[j] != 0 ){
                nums[i] = nums[j];
                i++;
            }
        }
        for(; i < nums.size(); i++){
            nums[i] = 0;
        }
    }
};

int main() {
    Solution s;
    vector<int> nums = {0, 1, 0, 3, 12};

    cout << "Before: ";
    for (int n : nums) cout << n << " ";
    cout << endl;

    s.moveZeroes(nums);

    cout << "After: ";
    for (int n : nums) cout << n << " ";
    cout << endl;

    return 0;
}

File: java/moveZeros.java
import java.util.Arrays;

public class moveZeros {

    public void moveZeroes(int[] nums) {
        int j = 0; // pointer for non-zero position

        for (int i = 0; i < nums.length; i++) {
            if (nums[i] != 0) {
                // Swap nums[i] with nums[j]
                int temp = nums[i];
                nums[i] = nums[j];
                nums[j] = temp;
                j++;
            }
        }
    }

    public static void main(String[] args) {
        moveZeros mz = new moveZeros();
        
        int[] nums = {0, 1, 0, 3, 12};
        System.out.println("Before: " + Arrays.toString(nums));
        
        mz.moveZeroes(nums);
        
        System.out.println("After:  " + Arrays.toString(nums));
    }
}
